gtfs-realtime/proto/gtfs-realtime.proto

```diff
@@ -399,9 +408,58 @@ message VehiclePosition {
   // including both seated and standing capacity, and current operating regulations allow.
   // It is possible that the value goes over 100 if there are currently more passengers than what the vehicle was designed for.
   // The precision of occupancy_percentage should be low enough that you can't track a single person boarding and alighting for privacy reasons.
+  // If multi_carriage_status is populated with per-carriage occupancy_percentage, 
+  // then this field should describe the entire vehicle with all carriages accepting passengers considered.
   // This field is still experimental, and subject to change. It may be formally adopted in the future.
   optional uint32 occupancy_percentage = 10;
 
+  // Carriage specific details, used for vehicles composed of several carriages
+  message CarriageStatus {
+
+    // User visible label that may be shown to the passenger to help identify
+    // the carriage. Example: "7712", "Car ABC-32", etc...
+    optional string label = 1;
+
+    // Occupancy status for this given carriage, in this vehicle
+    optional OccupancyStatus occupancy_status = 2 [default = NO_DATA_AVAILABLE];
+
+    // Occupancy percentage for this given carriage, in this vehicle
+    // Follows the same rules as "VehiclePosition.occupancy_percentage"
+    // -1 in case data is not available (as protobuf defaults to 0 otherwise)
+    optional int32 occupancy_percentage = 3 [default = -1];
```

Comment:
Unfortunately no - https://developers.google.com/protocol-buffers/docs/proto#optional says:

>When a message is parsed, if it does not contain an optional element, the corresponding field in the parsed object is set to the default value for that field. The default value can be specified as part of the message description...If the default value is not specified for an optional element, a type-specific default value is used instead: ...For numeric types, the default value is zero. For enums, the default value is the first value listed in the enum's type definition.

That means right now protobuf-generated client code is therefore interpreting `VehiclePosition.occupancy_percentage` as `0` if the producer hasn't populated the field, and `VehiclePosition.occupancy_percentage` is being interpreted as `EMPTY` if unpopulated.

To fix this `VehiclePosition.occupancy_percentage` needs to change from `uint32` to `int32` (which [should be](https://developers.google.com/protocol-buffers/docs/proto#updating) forwards/backwards compatible), with a default of -1 defined, and `OccupancyStatus` needs the additional `NO_DATA_AVAILABLE`, with `VehiclePosition.occupancy_percentage` using that as a default as well. Not defining these in the initial proposals for these fields was an oversight.

I plan to open a separate PR for this.

Reply:
@barbeau I sincerely hope uint32->int32 is also ABI compatible with end users.

Reply:
@skinkie That was my concern too. The link I referenced above, https://developers.google.com/protocol-buffers/docs/proto#updating, says:

>int32, uint32, int64, uint64, and bool are all compatible – this means you can change a field from one of these types to another without breaking forwards- or backwards-compatibility. If a number is parsed from the wire which doesn't fit in the corresponding type, you will get the same effect as if you had cast the number to that type in C++ (e.g. if a 64-bit number is read as an int32, it will be truncated to 32 bits).

Reply:
@barbeau it is wire compatible. I suppose this works for platforms with types like Python and Java. But for a C application this will be interesting. Now the occupancy percentage is something new, so I doubt there are many implementations.

Reply:
So thinking about this more, the main question would be what happens when a `-1` value that was serialized by a new `int32` producer is read by a legacy consumer still using the `uint32` in the .proto? I'll do some testing to see how this works.